src/hdf5_init_meta.c

```diff
@@ -0,0 +1,299 @@
+#include "hdf5_init_meta.h"
+#include "macros.h"
+
+#include <hdf5.h>
+#include <math.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+
+void create_meta_mq1_fields_dataset(hid_t *file_id,
+                                    hid_t *lcpl_id,
+                                    hid_t *meta_handle)
+{
+  if (meta_handle == NULL) {
+    fprintf(stderr, "meta_handle is NULL in create_meta_fields_dataset\n");
+    return;
+  }
+
+  hid_t file = *file_id;
+  hid_t lcpl = *lcpl_id;
+
+  char meta_group_path[64] = "metadata";
+  hid_t meta_group =
+    H5Gcreate(file, meta_group_path, lcpl, H5P_DEFAULT, H5P_DEFAULT);
+  if (meta_group < 0) {
+    fprintf(stderr, "Error creating group in create_meta_mq1_fields_dataset\n");
+    return;
+  }
+
+  hsize_t dim[1]       = {0};
+  hsize_t max_dim[1]   = {H5S_UNLIMITED};
+  hsize_t chunk_dim[1] = {1};
+
+  hid_t dataspace = H5Screate_simple(1, dim, max_dim);
+  if (dataspace < 0) {
+    fprintf(stderr,
+            "Error creating dataspace in create_meta_mq1_fields_dataset\n");
+    return;
+  }
+
+  hid_t dcpl = H5Pcreate(H5P_DATASET_CREATE);
+  if (dcpl < 0) {
+    fprintf(stderr, "Error creating dcpl in create_meta_mq1_fields_dataset\n");
+    H5Sclose(dataspace);
+    H5Gclose(meta_group);
+    return;
+  } else {
+    if (H5Pset_chunk(dcpl, 1, chunk_dim) < 0) {
+      fprintf(stderr,
+              "Error setting chunking in create_meta_mq1_fields_dataset\n");
+      H5Pclose(dcpl);
+      H5Sclose(dataspace);
+      H5Gclose(meta_group);
+      return;
+    }
+  }
+
+  hid_t dapl = H5Pcreate(H5P_DATASET_ACCESS);
+  if (dapl < 0) {
+    fprintf(stderr,
+            "Error in creating dapl in create_meta_mq1_fields_dataset\n");
+    H5Pclose(dcpl);
+    H5Sclose(dataspace);
+    H5Gclose(meta_group);
+    return;
+  } else {
+  }
+
+  // hid_t header_id_type = H5Tcopy(H5T_C_S1);
+  // H5Tset_size(header_id_type, MQ1_CHAR_LEN_HEADER_ID);
+
+  hid_t pixel_depth_type = H5Tcopy(H5T_C_S1);
+  H5Tset_size(pixel_depth_type, MQ1_CHAR_LEN_PIXEL_DEPTH);
+
+  hid_t sensor_layout_type = H5Tcopy(H5T_C_S1);
+  H5Tset_size(sensor_layout_type, MQ1_CHAR_LEN_SENSOR_LAYOUT);
+
+  hid_t chip_select_type = H5Tcopy(H5T_C_S1);
+  H5Tset_size(chip_select_type, MQ1_CHAR_LEN_CHIP_SELECT);
+
+  hid_t timestamp_type = H5Tcopy(H5T_C_S1);
+  H5Tset_size(timestamp_type, MQ1_CHAR_LEN_TIMESTAMP);
+
+  hid_t header_extension_id_type = H5Tcopy(H5T_C_S1);
+  H5Tset_size(header_extension_id_type, MQ1_CHAR_LEN_HEADER_EXTENSION_ID);
+
+  hid_t extended_timestamp_type = H5Tcopy(H5T_C_S1);
+  H5Tset_size(extended_timestamp_type, MQ1_CHAR_LEN_EXTENDED_TIMESTAMP);
+
+  hsize_t threshold_dims[1] = {MQ1_FLOAT_LEN_THRESHOLD};
+  hid_t threshold_type = H5Tarray_create(H5T_NATIVE_FLOAT, 1, threshold_dims);
```

Comment:
Will `H5Tcopy` or `H5Tarray_create` fail? If so, the resources need to be properly cleaned up like what you have done above.